Read and write LineBlock elements

tarleb opened the pull request:
This is work based on the second point of #1852 and the discussion in #1545.

The LineBlock element is converted to a Para with lines separated by hard line breaks by default.

The following writers have implementations leveraging appropriate format features:

 - AsciiDoc
 - ConTeXt
 - Custom
 - HTML
 - Markdown 
 - Org
 - RST

It's also supported by the following readers:

  - Docbook5
  - RST
  - Markdown
  - Org

The PR depends on jgm/pandoc-types#22.

ousia commented:
@tarleb, just in case it might help.

I guess a way to implement in the HTML writer would be to ad a `<div class="lineblock">` with the following code in the cascading stylesheet:

``` css
.lineblock{ white-space:pre-line; }
```

I’m not a coder myself, but I guess this may work.

tarleb commented:
Thanks for the input @ousia.  The problem with that approach is that all HTML fragments would have to contain a CSS block defining the `.lineblock` class.  People using pandoc without the `--standalone` flag would get the wrong layout.

ousia commented:
@tarleb, I thought of `<div style="white-space: pre-line;">`, but after that I realized that `<br/>` is a much simpler approach.

Sorry for the noise.

jgm commented:
You could alternatively add a literal style attribute, instead of
using a class that requires CSS.  This might actually be a
cleaner approach than using br tags.

+++ Albert Krewinkel [Aug 17 16 10:34 ]:
>   Thanks for the input [1]@ousia. The problem with that approach is that
>   all HTML fragments would have to contain a CSS block defining the
>   .lineblock class. People using pandoc without the --standalone flag
>   would get the wrong layout.
>
>   —
>   You are receiving this because you are subscribed to this thread.
>   Reply to this email directly, [2]view it on GitHub, or [3]mute the
>   thread.
>
>References
>
>   1. https://github.com/ousia
>   2. https://github.com/jgm/pandoc/pull/3072#issuecomment-240486892
>   3. https://github.com/notifications/unsubscribe-auth/AAAL5CEwlv-buyD9tLq9VKoOc6qvWQx9ks5qg0YogaJpZM4Jktv6

tarleb commented:
The writer methods for Org and ConTeXt currently have problems with overly long LineBlock lines.  I haven't yet found a simple way to solve that.

tarleb commented:
I learned about `nowrap` from Text.Pandoc.Pretty, so the issues mentioned in the previous comment have been resolved.

The behavior of the Markdown/RST lineblock parser was changed to always parse those line blocks into a single element.  Previously, if blocks contained blank lines, they were parsed into multiple paragraphs.

The Docbook5 writer no longer uses `linegroup`, which is only part of the 5.1 draft.

jgm commented:
Could you rebase these changes on top of master?
Ideally we'd smush these changes into two or three commits: perhaps one adding LineBlock support to the writers, one to the readers, and one for the changes in Shared.
The commit comments should explain the changes to the various readers and writers.
You should note explicitly the API change (exporting a new function from Shared).  That will help immensely when I write the changelog, otherwise I'm likely to forget to document that.

tarleb commented:
It seemed right to separate the change in Parsing from the other modifications as well, so there is a fourths commit on top of the other three.  I hope the commit-messages are suitable for the changelog in their present form.

Also, thanks again @ousia for the `pre-line` trick.

jgm commented on src/Text/Pandoc/Shared.hs:
```diff
@@ -630,6 +631,11 @@ compactify'DL items =
+-- | Combine a list of lines by adding hard linebreaks.  This is useful
+--   e.g. for rudimentary support of LineBlock elements in writers.
+combineLines :: [[Inline]] -> [Inline]
+combineLines = concat . map (++ [LineBreak])
```
Do we really want a LineBreak at the end of the last line?
Why not

    combineLines = intercalate [LineBreak]

and this is getting so short that maybe we don't even need a separate function?

tarleb replied:
My initial version did use `intercalate`, I do not quite remember why I changed it.  I think it was because rendering results were off in some formats (HTML), while it doesn't do any harm in other formats.  The switch in methods was my reason to introduce this function, exactly for the case where we'd want to change it again.

My current thoughts on it are that I'd rather add a function `linesToPara :: [[Inline]] -> Block` and export that instead.  This would save a tiny bit of boilerplate in the writers, as `combineLines` is almost exclusively used in combination with `Para`. Don't know why I didn't do this already.

jgm commented on src/Text/Pandoc/Parsing.hs:
```diff
@@ -718,11 +718,14 @@ lineBlockLine = try $ do
 lineBlockLines = try $ do
-  lines' <- many1 lineBlockLine
-  skipMany1 $ blankline <|> try (char '|' >> blankline)
+  lines' <- many1 (lineBlockLine <|> ((:[]) <$> blankLineBlockLine))
```
If we allow blank lines, we may run into problems in output formats that use paragraphs for line blocks and don't allow blank lines in paragraphs.  But maybe there aren't any such formats?  Just wanted to check.

tarleb replied:
AsciiDoc is the prime example here. This could be worked around, but changing the writer would probably take more time. I'll remote the commit from the PR.

jgm commented:
I looked at the pandoc manual; it doesn't say whether blank
lines are allowed in line blocks, but it does say that the
syntax is borrowed from reStructured Text.

Looking at reStructuredText,
http://docutils.sourceforge.net/docs/ref/rst/restructuredtext.html#line-blocks
I see that line blocks *can* contain blank lines.

So, I think it would actually be best to allow them to
contain blanks, and figure out how to work around this
in writers where this is a problem.

+++ Albert Krewinkel [Oct 11 16 11:29 ]:
>   @tarleb commented on this pull request.
>     __________________________________________________________________
>
>   In [1]src/Text/Pandoc/Parsing.hs:
>>  -- | Parses an RST-style line block and returns a list of strings.
> lineBlockLines :: Stream [Char] m Char => ParserT [Char] st m [String]
> lineBlockLines = try $ do
>-  lines' <- many1 lineBlockLine
>-  skipMany1 $ blankline <|> try (char '|' >> blankline)
>+  lines' <- many1 (lineBlockLine <|> ((:[]) <$> blankLineBlockLine))
>
>   AsciiDoc is the prime example here. This could be worked around, but
>   changing the writer would probably take more time. I'll remote the
>   commit from the PR.
>
>   —
>   You are receiving this because you commented.
>   Reply to this email directly, [2]view it on GitHub, or [3]mute the
>   thread.
>
>References
>
>   1. https://github.com/jgm/pandoc/pull/3072
>   2. https://github.com/jgm/pandoc/pull/3072
>   3. https://github.com/notifications/unsubscribe-auth/AAAL5K03LAZhqCLFkCjUT0jbapIAeUv4ks5qy9WQgaJpZM4Jktv6

jgm commented:
I see.  I think that line breaks at the end of paragraphs
are generally bad form, and may even be impossible in some
formats.

If HTML rendering looks bad, it may be a CSS issue.
It would help to have an example of what you mean, if you
can remember.

+++ Albert Krewinkel [Oct 11 16 10:49 ]:
>   @tarleb commented on this pull request.
>     __________________________________________________________________
>
>   In [1]src/Text/Pandoc/Shared.hs:
>> @@ -630,6 +631,11 @@ compactify'DL items =
>              | otherwise           -> items
>            _                       -> items
>
>+-- | Combine a list of lines by adding hard linebreaks.  This is useful
>+--   e.g. for rudimentary support of LineBlock elements in writers.
>+combineLines :: [[Inline]] -> [Inline]
>+combineLines = concat . map (++ [LineBreak])
>
>   My initial version did use intercalate, I do not quite remember why I
>   changed it. I think it was because rendering results were off in some
>   formats (HTML), while it doesn't do any harm in other formats. The
>   switch in methods was my reason to introduce this function, exactly for
>   the case where we'd want to change it again.
>
>   My current thoughts on it are that I'd rather add a function
>   linesToPara :: [[Inline]] -> Block and export that instead. This would
>   save a tiny bit of boilerplate in the writers, as combineLines is
>   almost exclusively used in combination with Para. Don't know why I
>   didn't do this already.
>
>   —
>   You are receiving this because you commented.
>   Reply to this email directly, [2]view it on GitHub, or [3]mute the
>   thread.
>
>References
>
>   1. https://github.com/jgm/pandoc/pull/3072
>   2. https://github.com/jgm/pandoc/pull/3072
>   3. https://github.com/notifications/unsubscribe-auth/AAAL5BBIXI40GjnqIMWEo4_FdsSEETkJks5qy8xDgaJpZM4Jktv6

tarleb commented:
Okay, fixed the AsciiDoc writer (the other writers seem to be working), changed `combineLines` to use `intercalate`, and rebased to master.  I haven't done extensive testing yet, but am fairly confident that things should work.

jgm commented:
Great work!
